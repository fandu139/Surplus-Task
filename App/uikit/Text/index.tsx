import StringUtils from '../../helper/string';
import Colors from '../../theme/colors';
import Fonts from '../../theme/fonts';
import React from 'react';
import { Text as RNText, StyleSheet, StyleProp, TextStyle, TextProps } from 'react-native';

interface Props extends TextProps {
  children: React.ReactNode;
  style?: StyleProp<TextStyle>;
  link?: boolean;
  error?: boolean;
  testID?: string;
  accessibilityLabel?: string;
}

interface TextComponent
  extends React.ForwardRefExoticComponent<Props & React.RefAttributes<RNText>> {
  Medium: React.ForwardRefExoticComponent<Props & React.RefAttributes<RNText>>;
  Bold: React.ForwardRefExoticComponent<Props & React.RefAttributes<RNText>>;
  Black: React.ForwardRefExoticComponent<Props & React.RefAttributes<RNText>>;
  Currency: React.ForwardRefExoticComponent<Props & React.RefAttributes<RNText>>;
}

const Text = React.forwardRef<RNText, Props>((props: Props, ref) => {
  const { children, style, link, error, testID, accessibilityLabel } = props;
  return (
    <RNText
      ref={ref}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...props}
      allowFontScaling={false}
      style={[styles.regularFamily, style, link && styles.link, error && styles.error]}
      accessibilityLabel={accessibilityLabel || testID}
    >
      {children}
    </RNText>
  );
}) as TextComponent;

const Medium = React.forwardRef((props: Props, ref: React.Ref<RNText>) => {
  const { children, style, link, error, testID, accessibilityLabel } = props;
  return (
    <RNText
      ref={ref}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...props}
      allowFontScaling={false}
      style={[styles.mediumFamily, style, link && styles.link, error && styles.error]}
      accessibilityLabel={accessibilityLabel || testID}
    >
      {children}
    </RNText>
  );
});

const Bold: React.ForwardRefExoticComponent<Props & React.RefAttributes<RNText>> = React.forwardRef(
  (props: Props, ref: React.Ref<RNText>) => {
    const { children, style, link, error, testID, accessibilityLabel } = props;
    return (
      <RNText
        ref={ref}
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...props}
        allowFontScaling={false}
        style={[styles.boldFamily, style, link && styles.link, error && styles.error]}
        accessibilityLabel={accessibilityLabel || testID}
      >
        {children}
      </RNText>
    );
  },
);

const Black = React.forwardRef((props: Props, ref: React.Ref<RNText>) => {
  const { children, style, link, error, testID, accessibilityLabel } = props;
  return (
    <RNText
      ref={ref}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...props}
      allowFontScaling={false}
      style={[styles.extraBoldFamily, style, link && styles.link, error && styles.error]}
      accessibilityLabel={accessibilityLabel || testID}
    >
      {children}
    </RNText>
  );
});

const Currency = React.forwardRef((props: Props, ref: React.Ref<RNText>) => {
  const { children, style, link, error, testID, accessibilityLabel } = props;
  return (
    <RNText
      ref={ref}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...props}
      allowFontScaling={false}
      style={[styles.regularFamily, style, link && styles.link, error && styles.error]}
      accessibilityLabel={accessibilityLabel || testID}
    >
      {`Rp ${StringUtils.thousandSeparator(children as string)}`}
    </RNText>
  );
});

const styles = StyleSheet.create({
  regularFamily: {
    ...Fonts.regular,
    color: Colors.BLACK,
    fontSize: Fonts.size.small,
  },
  mediumFamily: {
    ...Fonts.medium,
    color: Colors.BLACK,
    fontSize: Fonts.size.small,
  },
  boldFamily: {
    ...Fonts.bold,
    color: Colors.BLACK,
    fontSize: Fonts.size.small,
  },
  extraBoldFamily: {
    ...Fonts.extraBold,
    color: Colors.BLACK,
    fontSize: Fonts.size.small,
  },
  link: {
    color: Colors.BRIGHT_BLUE,
  },
  error: {
    color: Colors.RED,
  },
});

Text.defaultProps = {
  link: false,
  style: {},
  error: false,
};

Text.Bold = Bold;
Text.Medium = Medium;
Text.Currency = Currency;
Text.Black = Black;

export default Text;
