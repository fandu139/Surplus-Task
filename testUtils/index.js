import AuthContext from '../App/context/Auth';
import React from 'react';
import { act, fireEvent, render } from '@testing-library/react-native';

const customRender = (ui, options) => {
  const isAuthenticated = options?.isAuthenticated || false;
  const renderOptions = options?.rendererOptions;
  const authContext = options?.authContext;
  return render(
    <AuthContext.Provider value={{ isAuthenticated, ...authContext }}>{ui}</AuthContext.Provider>,
    renderOptions,
  );
};

// override render method
export { customRender as render };

export function wait(ms) {
  return act(() => new Promise((r) => setTimeout(r, ms)));
}

// Due to debounce on Button UI Kit, we need to add milisecond waiting timer
export function fireAsyncPress(element, ms = 100) {
  fireEvent.press(element);
  return wait(ms);
}
