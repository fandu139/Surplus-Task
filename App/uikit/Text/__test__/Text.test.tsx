import Text from '../';
import React from 'react';
import { render } from '../../../../testUtils';

describe('Text component test', () => {
  describe('Text snapshot test', () => {
    it('should render correctly', () => {
      const tree = render(
        <Text link error>
          Test Text
        </Text>,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Text Medium snapshot test', () => {
    it('should render correctly', () => {
      const tree = render(
        <Text.Medium link error>
          Test Text
        </Text.Medium>,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Text Black snapshot test', () => {
    it('should render correctly', () => {
      const tree = render(
        <Text.Black link error>
          Test Text
        </Text.Black>,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Text Bold snapshot test', () => {
    it('should render correctly', () => {
      const tree = render(
        <Text.Bold link error>
          Test Text
        </Text.Bold>,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Text Currency snapshot test', () => {
    it('should render correctly', () => {
      const tree = render(
        <Text.Currency link error>
          Test Text
        </Text.Currency>,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
