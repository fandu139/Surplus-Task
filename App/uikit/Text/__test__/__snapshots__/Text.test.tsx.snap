// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text component test Text Black snapshot test should render correctly 1`] = `
<Text
  allowFontScaling={false}
  error={true}
  link={true}
  style={
    [
      {
        "color": "#202020",
        "fontSize": 14,
        "fontWeight": "normal",
      },
      undefined,
      {
        "color": "#00A2D8",
      },
      {
        "color": "#EF4941",
      },
    ]
  }
>
  Test Text
</Text>
`;

exports[`Text component test Text Bold snapshot test should render correctly 1`] = `
<Text
  allowFontScaling={false}
  error={true}
  link={true}
  style={
    [
      {
        "color": "#202020",
        "fontSize": 14,
        "fontWeight": "normal",
      },
      undefined,
      {
        "color": "#00A2D8",
      },
      {
        "color": "#EF4941",
      },
    ]
  }
>
  Test Text
</Text>
`;

exports[`Text component test Text Currency snapshot test should render correctly 1`] = `
<Text
  allowFontScaling={false}
  error={true}
  link={true}
  style={
    [
      {
        "color": "#202020",
        "fontSize": 14,
        "fontWeight": "normal",
      },
      undefined,
      {
        "color": "#00A2D8",
      },
      {
        "color": "#EF4941",
      },
    ]
  }
>
  Rp 0
</Text>
`;

exports[`Text component test Text Medium snapshot test should render correctly 1`] = `
<Text
  allowFontScaling={false}
  error={true}
  link={true}
  style={
    [
      {
        "color": "#202020",
        "fontSize": 14,
        "fontWeight": "normal",
      },
      undefined,
      {
        "color": "#00A2D8",
      },
      {
        "color": "#EF4941",
      },
    ]
  }
>
  Test Text
</Text>
`;

exports[`Text component test Text snapshot test should render correctly 1`] = `
<Text
  allowFontScaling={false}
  error={true}
  link={true}
  style={
    [
      {
        "color": "#202020",
        "fontSize": 14,
        "fontWeight": "normal",
      },
      {},
      {
        "color": "#00A2D8",
      },
      {
        "color": "#EF4941",
      },
    ]
  }
>
  Test Text
</Text>
`;
